boltdb/bolt_store.go

```diff
@@ -3,413 +3,169 @@ package boltdb
-			pre := append([]byte(key), 0)
-			for k, v := c.Seek(pre); bytes.HasPrefix(k, pre); k, v = c.Next() {
-				send := false
-				if filter != nil {
-					e := aql.Edge{}
-					proto.Unmarshal(v, &e)
-					if filter(e) {
-						send = true
-					}
-				} else {
-					send = true
-				}
-				if send {
-					pair := bytes.Split(k, []byte{0})
-					vo <- string(pair[1])
-				}
-			}
-			return nil
-		})
-	}()
-	go func() {
-		defer close(o)
-		for i := range vo {
-			if loadProp {
-				v := self.GetVertex(i, loadProp)
-				if v == nil {
-					//log.Printf("Vertex Missing %s", i)
-				} else {
-					o <- *v
-				}
-			} else {
-				o <- aql.Vertex{Gid: i}
-			}
-		}
-	}()
-	return o
+func (self *BoltKV) Update(u func(tx kvgraph.KVTransaction) error) error {
+	err := self.db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket(GRAPH_BUCKET)
+		ktx := BoltTransaction{tx, b}
+		return u(ktx)
+	})
+	return err
 }
 
-func (self *BoltArachne) GetInList(key string, loadProp bool, filter gdbi.EdgeFilter) chan aql.Vertex {
-	vi := make(chan string, 100)
-	o := make(chan aql.Vertex, 100)
-	go func() {
-		defer close(vi)
-		self.db.View(func(tx *bolt.Tx) error {
-			eb := tx.Bucket(IEdgeBucket)
-			ob := tx.Bucket(OEdgeBucket)
-			c := eb.Cursor()
-			pre := append([]byte(key), 0)
-			for k, _ := c.Seek(pre); bytes.HasPrefix(k, pre); k, _ = c.Next() {
-				key_data := bytes.Split(k, []byte{0})
-				send := false
-				if filter != nil {
-					e := aql.Edge{}
-					ikey := bytes.Join([][]byte{[]byte(key_data[1]), []byte(key_data[0]), []byte(key_data[2])}, []byte{0})
-					d := ob.Get([]byte(ikey))
-					proto.Unmarshal(d, &e)
-					if filter(e) {
-						send = true
-					}
-				} else {
-					send = true
-				}
-				if send {
-					vi <- string(key_data[1])
-				}
-			}
-			return nil
-		})
-	}()
-	go func() {
-		defer close(o)
-		for i := range vi {
-			if loadProp {
-				o <- *self.GetVertex(i, loadProp)
-			} else {
-				o <- aql.Vertex{Gid: i}
-			}
-		}
-	}()
-	return o
+type BoltIterator struct {
+	tx    *bolt.Tx
+	b     *bolt.Bucket
+	c     *bolt.Cursor
+	key   []byte
+	value []byte
 }
 
-func (self *BoltArachne) GetOutEdgeList(key string, loadProp bool, filter gdbi.EdgeFilter) chan aql.Edge {
-	o := make(chan aql.Edge, 100)
-	go func() {
-		defer close(o)
-		self.db.View(func(tx *bolt.Tx) error {
-			eb := tx.Bucket(OEdgeBucket)
-			c := eb.Cursor()
-			pre := append([]byte(key), 0)
-			for k, v := c.Seek(pre); bytes.HasPrefix(k, pre); k, v = c.Next() {
-				send := false
-				e := aql.Edge{}
-				if loadProp {
-					proto.Unmarshal(v, &e)
-				} else {
-					key_data := bytes.Split(k, []byte{0})
-					e.Out = string(key_data[0])
-					e.In = string(key_data[1])
-					e.Gid = string(key_data[2])
-				}
-				if filter != nil {
-					if filter(e) {
-						send = true
-					}
-				} else {
-					send = true
-				}
-				if send {
-					o <- e
-				}
-			}
-			return nil
-		})
-	}()
-	return o
+func bytes_copy(in []byte) []byte {
+	out := make([]byte, len(in))
```

Comment:
Add the capacity as well, `make([]byte, len(in), len(in))`

Reply:
If you use two arguments, the capacity is the length

Reply:
Ah, I didn't know that. Seems obvious now. Thanks.